### PR TITLE
Refactor IsAutoPropertyWithGetAccessor usage in MethodCompiler

### PR DESCRIPTION
**Pull request by Youssef1313**
@AlekseyTs This might not be what you're expecting. It's not much clear to me what "clean up" is needed.

Test plan: https://github.com/dotnet/roslyn/issues/57012
Proposal: https://github.com/dotnet/csharplang/issues/140

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 1]
```diff
@@ -1684,7 +1684,7 @@ private static bool AccessingAutoPropertyFromConstructor(BoundExpression receive
             var propertyIsStatic = propertySymbol.IsStatic;
 
             return (object)sourceProperty != null &&
-                    sourceProperty.IsAutoPropertyWithGetAccessor &&
+                    sourceProperty.IsAutoProperty && sourceProperty.GetMethod is not null &&
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:1687]: This will need future adjustment specific to semi auto properties. I didn't add a PROTOTYPE here since there is a one in the tests file already (`AssignReadOnlyOnlyPropertyInConstructor`)

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 2]
```diff
@@ -1870,7 +1870,7 @@ internal static BoundBlock BindMethodBody(MethodSymbol method, TypeCompilationSt
                 else
                 {
                     var property = sourceMethod.AssociatedSymbol as SourcePropertySymbolBase;
-                    if ((object)property != null && property.IsAutoPropertyWithGetAccessor) // PROTOTYPE: Just IsAutoProperty, or _isAutoPropertyAccessor?
+                    if ((object)property != null && property.IsAutoProperty && property.GetMethod is not null) // PROTOTYPE: Just IsAutoProperty, or _isAutoPropertyAccessor?
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:1873]: TThis will need future adjustment for mixed scenario. Here we should be caring about whether the accessor is a semi-colon only accessor or not. In fact, this code path shouldn't be reachable for non-semi-colon only accessor. So an assert will probably be needed.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:1873]: >property.IsAutoProperty && property.GetMethod is not null [](http://example.com/codeflow?start=52&length=57)

If I remember correctly, in one of the comments on the initial PR I indicated that we simply shouldn't be asking property about anything here. An accessor should itself know if its body is supposed to be synthesized. This is the kind of refactoring I had in mind for this place.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:1873]: > This will need future adjustment for mixed scenario. Here we should be caring about whether the accessor is a semi-colon only accessor or not. In fact, this code path shouldn't be reachable for non-semi-colon only accessor. So an assert will probably be needed.

A simple in-lining of implementation isn't the kind of refactoring I had in mind

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs [hunk 3]
```diff
@@ -301,7 +301,7 @@ private BoundExpression VisitAssignmentOperator(BoundAssignmentOperator node, bo
             if (setMethod is null)
             {
                 var autoProp = (SourcePropertySymbolBase)property.OriginalDefinition;
-                Debug.Assert(autoProp.IsAutoPropertyWithGetAccessor,
+                Debug.Assert(autoProp.IsAutoProperty && autoProp.GetMethod is not null,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs:304]: >autoProp.IsAutoProperty && autoProp.GetMethod is not null [](http://example.com/codeflow?start=29&length=57)

I think you are missing the point of the activity that we are doing. A simple in-lining of implementation isn't the kind of refactoring I had in mind. Think what is the purpose of this check, i.e. what are we trying actually to check. Then think if it is worth adding a new property that has that specific purpose in mind. If you cannot come up with anything, skip the usage.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:1687]: >sourceProperty.IsAutoProperty && sourceProperty.GetMethod is not null [](http://example.com/codeflow?start=20&length=69)

Same comment as for other places. Without taking a deep look, here is the theory that I have: "Here we actually want to check if an assignment to a property within constructor is supposed to assign to a backing field instead." If my theory is generally correct, let's add a new property with the specific name and purpose, and let's use it here. It would feel appropriate to rename the containing method accordingly as well. This is the process that we should go through for each API usage, one by one.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs [hunk 4]
```diff
@@ -450,7 +449,7 @@ private void EnsureBackingFieldIsSynthesized()
             }
 
             // If binding both the getter and setter didn't get a backing field, set it to null so that we don't re-calculate.
-            Interlocked.CompareExchange(ref _lazyBackingFieldSymbol, null, _lazyBackingFieldSymbolSentinel);
+            MarkBackingFieldAsCalculated();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs:452]: >MarkBackingFieldAsCalculated [](http://example.com/codeflow?start=12&length=28)

How is this change related to the goal of this PR. Please keep things separate. If you would like to make this refactoring, please submit a separate PR.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs [hunk 5]
```diff
@@ -828,10 +824,8 @@ internal override void AfterAddingTypeMembersChecks(ConversionsBase conversions,
                 CheckInitializer(IsAutoProperty, ContainingType.IsInterface, IsStatic, Location, diagnostics);
             }
 
-            if (IsAutoPropertyWithGetAccessor)
+            if (IsAutoProperty && GetMethod is not null)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs:827]: >IsAutoProperty && GetMethod is not null [](http://example.com/codeflow?start=16&length=39)

Similar comment

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs [hunk 6]
```diff
@@ -1197,7 +1191,7 @@ private SynthesizedSealedPropertyAccessor MakeSynthesizedSealedAccessor()
         AttributeLocation IAttributeTargetSymbol.DefaultAttributeLocation => AttributeLocation.Property;
 
         AttributeLocation IAttributeTargetSymbol.AllowedAttributeLocations
-            => IsAutoPropertyWithGetAccessor // PROTOTYPE(semi-auto-props): Adjust this and add tests.
+            => IsAutoProperty && GetMethod is not null // PROTOTYPE(semi-auto-props): Adjust this and add tests.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs:1194]: >IsAutoProperty && GetMethod is not null [](http://example.com/codeflow?start=15&length=39)

Similar comment as for other inlining.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs [hunk 7]
```diff
@@ -1648,7 +1642,7 @@ protected virtual void ValidatePropertyType(BindingDiagnosticBag diagnostics)
             {
                 diagnostics.Add(ErrorCode.ERR_FieldCantBeRefAny, TypeLocation, type);
             }
-            else if (this.IsAutoPropertyWithGetAccessor && type.IsRefLikeType && (this.IsStatic || !this.ContainingType.IsRefLikeType))
+            else if (this.IsAutoProperty && type.IsRefLikeType && (this.IsStatic || !this.ContainingType.IsRefLikeType))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs:1645]: >this.IsAutoProperty [](http://example.com/codeflow?start=21&length=19)

The motivation behind this change is not clear to me. Could you please elaborate?

File: src/Compilers/CSharp/Test/Emit/Emit/EmitMetadataTests.cs [hunk 8]
```diff
@@ -1293,7 +1293,8 @@ private static void VerifyAutoProperty(PropertySymbol property, bool isFromSourc
             {
                 if (property is SourcePropertySymbol sourceProperty)
                 {
-                    Assert.True(sourceProperty.IsAutoPropertyWithGetAccessor);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/EmitMetadataTests.cs:1296]: >Assert.True(sourceProperty.IsAutoPropertyWithGetAccessor); [](http://example.com/codeflow?start=20&length=58)

Inlining implementation in test might be Ok, but I would wait until after that is the only usage of the ```IsAutoPropertyWithGetAccessor``` property.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedBackingFieldSymbol.cs [hunk 9]
```diff
@@ -134,7 +158,7 @@ internal override void PostDecodeWellKnownAttributes(ImmutableArray<CSharpAttrib
         {
             base.PostDecodeWellKnownAttributes(boundAttributes, allAttributeSyntaxNodes, diagnostics, symbolPart, decodedData);
 
-            if (!allAttributeSyntaxNodes.IsEmpty && _property.IsAutoPropertyWithGetAccessor)
+            if (!allAttributeSyntaxNodes.IsEmpty && _property.IsAutoProperty && _property.GetMethod is not null)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedBackingFieldSymbol.cs:161]: >_property.IsAutoProperty && _property.GetMethod is not null [](http://example.com/codeflow?start=52&length=59)

Similar comment as for other inlining.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedBackingFieldSymbol.cs [hunk 10]
```diff
@@ -18,11 +19,22 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
...
+            IsCreatedForFieldKeyword = 1 << 1,
+            IsEarlyConstructed = 1 << 2,
+        }
+
         private readonly SourcePropertySymbolBase _property;
         private readonly string _name;
-        internal bool HasInitializer { get; }
-        internal bool IsCreatedForFieldKeyword { get; }
-        internal bool IsEarlyConstructed { get; }
+        private readonly Flags _backingFieldFlags;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedBackingFieldSymbol.cs:32]: >private readonly Flags _backingFieldFlags; [](http://example.com/codeflow?start=8&length=42)

It looks like this refactoring is not related to the goal of this PR. Please extract into a separate PR.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs (hunk 4 above)
Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs:452]: @AlekseyTs This was addressing @333fred feedback that was asked to be done in the next PR. I'm okay with addressing the feedback separately.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs:452]: > This was addressing feedback that was asked to be done in the next PR. I'm okay with addressing the feedback separately.

The definition of the next PR is quite vague. Is it the next submitted PR, or is it the next merged PR? These are different things. So, in general, we understand that the term "next PR" refers to a PR in a **near future** that addresses the concern.

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs (hunk 2 above)
Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:1873]: @AlekseyTs I refactored this a bit, but I ended up that I'll need to ask the property whether it has a get accessor or not, otherwise we'll call `ConstructAutoPropertyAccessorBody` and crash because there is no backing field for a semi-colon setter only property (e.g `public int P { set; }`).

We can prevent the crash by creating a backing field in absence of get accessor, but I don't think we really want to unnecessarily create a backing field, which also (may) affect other parts of API consumers.

What do you think?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:1873]: >  I refactored this a bit, but I ended up that I'll need to ask the property whether it has a get accessor or not

This sounds wrong. If accessor knows whether its body is supposed to be synthesized and that is something that is known at its construction time, then why do we need to ask property anything here?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:1873]: Here is the relevant thread from the initial PR - [https://github.com/dotnet/roslyn/pull/57076/commits/82b632a0c8b57117ff60f8c09c08b3f599b9aae4#r773483332](https://github.com/dotnet/roslyn/pull/57076/commits/82b632a0c8b57117ff60f8c09c08b3f599b9aae4#r773483332)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs [hunk 11]
```diff
@@ -213,7 +209,7 @@ internal sealed override ImmutableArray<string> NotNullWhenFalseMembers
                    isIterator)
         {
             _property = property;
-            _isAutoPropertyAccessor = isAutoPropertyAccessor;
+            _accessorBodyShouldBeSynthesized = property.IsAutoProperty && syntax is AccessorDeclarationSyntax { SemicolonToken.RawKind: (int)SyntaxKind.SemicolonToken, ExpressionBody: null, Body: null };
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:212]: @AlekseyTs I'm still asking the property a question, but the question has method from `MethodCompiler` to `SourcePropertyAccessorSymbol`. Is that the kind of refactor you want? I can't think of a way to completely avoid asking the property.

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:212]: Actually I think this will crash, should have used `IsAutoPropertyWithGetAccessor`, or the other solution is to handle null backing field in `ConstructAutoPropertyAccessorBody`. Let me know what you prefer

**Comment by Youssef1313** [comment]
@AlekseyTs I reverted all the simple inlinings for now.

File: src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.cs [hunk 12]
```diff
@@ -184,6 +184,13 @@ internal static BoundBlock ConstructAutoPropertyAccessorBody(SourceMemberMethodS
             }
 
             var field = property.BackingField;
+            if (field is null)
+            {
+                // This happens for public int { set; } where we produce ERR_AutoPropertyMustHaveGetAccessor
+                Debug.Assert(!property.HasGetAccessor);
+                return null;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.cs:191]: >return null; [](http://example.com/codeflow?start=16&length=12)

Consider adjusting the doc comment to indicate that null can be returned in case of an error.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEqualityContractProperty.cs [hunk 13]
```diff
@@ -55,7 +55,7 @@ public override SyntaxList<AttributeListSyntax> AttributeDeclarationSyntaxList
         protected override Location TypeLocation
             => ContainingType.Locations[0];
 
-        protected override SourcePropertyAccessorSymbol CreateGetAccessorSymbol(bool isAutoPropertyAccessor, BindingDiagnosticBag diagnostics)
+        protected override SourcePropertyAccessorSymbol CreateGetAccessorSymbol(BindingDiagnosticBag diagnostics)
         {
             return SourcePropertyAccessorSymbol.CreateAccessorSymbol(
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEqualityContractProperty.cs:60]: >CreateAccessorSymbol [](http://example.com/codeflow?start=48&length=20)

It looks like this is the only call site for this method and its implementation is very specialized for this class. I think it would be good to remove it and simply inline its body here. After that we should be able to make ```GetAccessorSymbol``` private.

**Comment by AlekseyTs** [comment]
>             this.CheckModifiers(location, hasBody: true, isAutoPropertyOrExpressionBodied: true, diagnostics: diagnostics);

While we are here, let's find a better name for this parameter. Something that doesn't use the "AutoProperty" term which becomes confusing while we are transitioning from its old meaning to the new meaning.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:188 in 6d53a72. [](commit_id = 6d53a72250dbb3d9387cba34a9c1737f62e4ac34, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>             this.CheckModifiers(location, hasBody: true, isAutoPropertyOrExpressionBodied: true, diagnostics: diagnostics);

It looks like we can remove this parameter completely, ```hasBody``` should be sufficient. 

---
In reply to: [1002657565](https://github.com/dotnet/roslyn/pull/58515#issuecomment-1002657565) [](http://example.com/codeflow?ancestors=1002657565)

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:188 in 6d53a72. [](commit_id = 6d53a72250dbb3d9387cba34a9c1737f62e4ac34, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs [hunk 14]
```diff
@@ -254,7 +251,7 @@ internal sealed override ImmutableArray<string> NotNullWhenFalseMembers
 
             if (!modifierErrors)
             {
-                this.CheckModifiers(location, hasBody || hasExpressionBody, isAutoPropertyAccessor, diagnostics);
+                this.CheckModifiers(location, hasBody || hasExpressionBody, _accessorBodyShouldBeSynthesized, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:254]: It looks like this can be folded into the previous parameter value and this parameter can be removed.

**Comment by AlekseyTs** [comment]
>             bool hasBody,

Would it make sense to rename this to "hasBlockBody"?

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:198 in 6d53a72. [](commit_id = 6d53a72250dbb3d9387cba34a9c1737f62e4ac34, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.cs [hunk 15]
```diff
@@ -184,6 +184,13 @@ internal static BoundBlock ConstructAutoPropertyAccessorBody(SourceMemberMethodS
             }
 
             var field = property.BackingField;
+            if (field is null)
+            {
+                // This happens for public int { set; } where we produce ERR_AutoPropertyMustHaveGetAccessor
+                Debug.Assert(!property.HasGetAccessor);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.cs:190]: >property.HasGetAccessor [](http://example.com/codeflow?start=30&length=23)

There is an existing API to check for a setter. We don't need to expose a new one.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs [hunk 16]
```diff
@@ -690,7 +686,7 @@ public sealed override ImmutableArray<ParameterSymbol> Parameters
         internal override bool IsExplicitInterfaceImplementation
             => (_propertyFlags & Flags.IsExplicitInterfaceImplementation) != 0;
 
-        private bool HasGetAccessor => (_propertyFlags & Flags.HasGetAccessor) != 0;
+        internal bool HasGetAccessor => (_propertyFlags & Flags.HasGetAccessor) != 0;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs:689]: >internal [](http://example.com/codeflow?start=8&length=8)

There is an existing API to check for a setter. We don't need to expose a new one.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs [hunk 17]
```diff
@@ -478,6 +475,11 @@ internal Accessibility LocalAccessibility
         /// </remarks>
         internal bool ContainsFieldKeyword => _containsFieldKeyword;
 
+        /// <summary>
+        /// Indicates whether this accessor has no body in source and a body will be synthesized by the compiler.
+        /// </summary>
+        internal bool AccessorBodyShouldBeSynthesized => _accessorBodyShouldBeSynthesized;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:481]: >Accessor [](http://example.com/codeflow?start=22&length=8)

I think we can drop the "Accessor" prefix from the name of the property and the backing field.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs [hunk 18]
```diff
@@ -213,7 +209,8 @@ internal sealed override ImmutableArray<string> NotNullWhenFalseMembers
                    isIterator)
         {
             _property = property;
-            _isAutoPropertyAccessor = isAutoPropertyAccessor;
+            _accessorBodyShouldBeSynthesized = (property.IsAutoProperty && syntax is AccessorDeclarationSyntax { SemicolonToken.RawKind: (int)SyntaxKind.SemicolonToken, ExpressionBody: null, Body: null }) ||
+                                                property is SynthesizedRecordPropertySymbol;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:213]: >property is SynthesizedRecordPropertySymbol [](http://example.com/codeflow?start=48&length=43)

Instead of handling all possible flavors of accessors here, I think it would be better for the consumer to pass ```bodyShouldBeSynthesized``` flag to use (for record accessors we don't even need to try to check syntax). Basically we will have three body-related flags: ```hasBody```, ```hasExpressionBody``` and ```bodyShouldBeSynthesized```. Also, I think we should avoid using ```IsAutoProperty```. Passing a flag, let's say named something like "bodyShouldBeSynthesizedForSemicolonOnly", from property's constructor should be sufficient.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs [hunk 19]
```diff
@@ -194,11 +194,11 @@ private enum Flags : byte
 
             if (hasGetAccessor)
             {
-                _getMethod = CreateGetAccessorSymbol(isAutoPropertyAccessor: isAutoProperty, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs:197]: >isAutoPropertyAccessor: isAutoProperty,  [](http://example.com/codeflow?start=53&length=40)

I think we should keep passing ```isAutoProperty``` value, but its corresponding parameter should change its name and its meaning. The "bodyShouldBeSynthesizedForSemicolonOnly" name feels like a good fit.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs [hunk 20]
```diff
@@ -194,11 +194,11 @@ private enum Flags : byte
...
+                _getMethod = CreateGetAccessorSymbol(diagnostics);
             }
             if (hasSetAccessor)
             {
-                _setMethod = CreateSetAccessorSymbol(isAutoPropertyAccessor: isAutoProperty, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs:201]: >isAutoProperty [](http://example.com/codeflow?start=77&length=14)

Similar comment here.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs [hunk 21]
```diff
@@ -213,7 +209,8 @@ internal sealed override ImmutableArray<string> NotNullWhenFalseMembers
                    isIterator)
         {
             _property = property;
-            _isAutoPropertyAccessor = isAutoPropertyAccessor;
+            _accessorBodyShouldBeSynthesized = (property.IsAutoProperty && syntax is AccessorDeclarationSyntax { SemicolonToken.RawKind: (int)SyntaxKind.SemicolonToken, ExpressionBody: null, Body: null }) ||
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:212]: >SemicolonToken.RawKind: (int)SyntaxKind.SemicolonToken [](http://example.com/codeflow?start=113&length=54)

It doesn't look like we actually need to check anything for the semicolon token itself. ```SourcePropertySymbol.GetAccessorDeclarations``` only checks if we have any explicit body. Let's be consistent and avoid any possible deviations around error scenarios.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 9)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs [hunk 22]
```diff
@@ -58,7 +58,7 @@ internal class SourcePropertyAccessorSymbol : SourceMemberMethodSymbol
                 syntax.Modifiers,
                 methodKind,
                 syntax.Keyword.IsKind(SyntaxKind.InitKeyword),
-                isAutoPropertyAccessor,
+                bodyShouldBeSynthesized: bodyShouldBeSynthesizedForSemicolonOnly && !hasBody && !hasExpressionBody,
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:61]: @AlekseyTs I'm not sure if this is the kind of thing you wanted.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs [hunk 23]
```diff
@@ -105,28 +105,11 @@ internal class SourcePropertyAccessorSymbol : SourceMemberMethodSymbol
                 modifiers: new SyntaxTokenList(),
                 methodKind,
                 usesInit,
-                isAutoPropertyAccessor: true,
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:108]: I think we could create a new class, `SynthesizedRecordPropertyAccessorSymbol`, which could be an inner private class to `SynthesizedRecordPropertySymbol`, and remove this method from here. @AlekseyTs Does this suggestion look good? Should I go for it in this PR or a separate PR?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:108]: >I think we could create a new class, ```SynthesizedRecordPropertyAccessorSymbol``, which could be an inner private class to ```SynthesizedRecordPropertySymbol```, and remove this method from here. Does this suggestion look good? Should I go for it in this PR or a separate PR?

What benefits do you see from introducing that new class? Will it have special behavior for some scenarios? Would simply moving the method to ``````SynthesizedRecordPropertySymbol``` work?

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:108]: @AlekseyTs Moving the method will require exposing some constructor.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:108]: I wouldn't introduce a new class just to move the method.

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:108]: So should I move it and make the constructor `internal` or keep it as it is?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs:108]: > So should I move it and make the constructor ``internal``` or keep it as it is?

Let's keep it as is. There is really not much difference if we move it. We will still need a special entry point here, the exposed constructor, and letting everyone access it doesn't feel like a change to the better.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs (hunk 22 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: > I'm not sure if this is the kind of thing you wanted.

Yes, looks good. Thanks.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 12)

**Comment by AlekseyTs** [comment]
@333fred, @dotnet/roslyn-compiler For the second review.

**Comment by Youssef1313** [comment]
@AlekseyTs Is there any more refactoring needed?

If not, then the next step is looking at supporting mixed scenarios right? (https://github.com/dotnet/roslyn/pull/57076#issuecomment-999645691 for reference)

**Comment by AlekseyTs** [comment]
> Is there any more refactoring needed before continuing working on the feature?

I think there is. We need to examine other usages of "AutoProperty" related APIs and introduce suitable abstractions that can be used instead and were not about the "AutoProperty" term. I gave you an example here https://github.com/dotnet/roslyn/pull/58515#discussion_r775999953. The theory there is probably not quite accurate, but it should be possible to get the idea. Let's start with that specific place and go from there.

Also, before we start working on any behavior changes, I would like us to get earlier identified tests in place. In particular those about ```SemanticModel``` and ```GetMembers``` API.

**Comment by AlekseyTs** [comment]
The link to the comment doesn't quite work. It was about a change in earlier iteration of this PR in ```AccessingAutoPropertyFromConstructor``` in BinderStatements.cs

Here is the interesting part: 
>Without taking a deep look, here is the theory that I have: "Here we actually want to check if an assignment to a property within constructor is supposed to assign to a backing field instead." If my theory is generally correct, let's add a new property with the specific name and purpose, and let's use it here. It would feel appropriate to rename the containing method accordingly as well. This is the process that we should go through for each API usage, one by one.

Again, without significant additional thinking, I have an idea that it might help to add a couple of properties to SourcePropertySymbolBase instead. Something like:
- GetMethodIsEquivalentToBackingFieldRead
- SetMethodIsEquivalentToBackingFieldWrite

Perhaps the ```AccessingAutoPropertyFromConstructor``` should become more specific and probably be split between Get and Set access. 

Just think about what you can come up with. The goal is to introduce abstractions that do not refer to "AutoProperty" and at the same time likely to have an application for semi-auto properties, or have no relation to them at all.

**Comment by Youssef1313** [comment]
> * GetMethodIsEquivalentToBackingFieldRead
> * SetMethodIsEquivalentToBackingFieldWrite

So `[Get|Set]MethodIsEquivalentToBackingField[Read|Write]` can only be true if the getter/setter is semi-colon only right? It's essentially the same meaning as "Is this a semicolon-only accessor of an auto property", or am I missing something?

**Comment by AlekseyTs** [comment]
>So `[Get|Set]MethodIsEquivalentToBackingField[Read|Write]` can only be true if the getter/setter is semi-colon only right? It's essentially the same meaning as "Is this a semicolon-only accessor of an auto property", or am I missing something?

Pretty much. However, it would be good to move away from using "an auto property" term even in implementation, when possible. For example, the implementation for properties could be as simple as a check whether ```BodyShouldBeSynthesized``` is true for the corresponding accessor, I think.

**Comment by AlekseyTs** [comment]
@333fred, @dotnet/roslyn-compiler For the second review.

**Comment by AlekseyTs** [comment]
@333fred, @dotnet/roslyn-compiler For the second review of a community PR.

**Review by 333fred — APPROVED** [review]
LGTM (commit 12)
